### ISSUE TITLE
More consistent support for single precision computation across skimage

### ISSUE THREAD

**Issue by grlee77** [scikit-image/scikit-image#5205]
## Description

Some functions in scikit-image support single precision and many can potentially do so with minimal modification. I opened this issue to track PRs related to float32 support and discuss best practices to make behavior more uniform across functions.

There are corner cases such as non-local means using the integral image approach where accuracy considerations require internal use of float64, but in most cases float32 precision is acceptable.

### API considerations

Some considerations are:
   - should we infer floating precision based on input dtype?
   - When inferring floating point dtype for integer inputs, should we use the result of `np.promote_types(image.dtype, np.float32)` or always use `np.float64`? `promote_types` promotes 8 and 16-bit integers to `float32`, but larger integers to float64.
   - should we generally provide a dtype argument to allow the user to override the default?

We have some functions such as `optical_flow_ilk` and `optical_flow_tvl1` that use a `dtype` argument to control the precision. For those two it defaults to `dtype=float32`.

I am generally in favor of preserving the floating point precision of the input (although usually promoting float16 to float32). This is what I implemented for `phase_cross_correlation` and `masked_phase_cross_correlation` in #5204.

PR #4307 adds float32 support to `moments_hu` and infers precision from the input for the default `dtype=None`, but also allows overriding this default.

### open issues and PRs related to dtype handling

#4307: moments_hu
#5204: phase_cross_correlation, masked_phase_cross_correlation

**Comment by grlee77** [comment]
On the questions above, I generally prefer inferring the dtype of the operation based on the input and would not generally encourage an additional dtype argument.

I don't have a strong opinion on whether integer inputs should always be cast to double. It might be safest from a backwards compatibility standpoint?

**Comment by FirefoxMetzger** [comment]
A potentially stupid question: What is the benefit of using `np.float32` over `np.float64` for computation?

**Comment by rfezzani** [comment]
It is absolutely not a stupid question. Single precision float requires half storage memory then double precision and operations are usually faster. That's particularly true on some hardware devices (like GPUs)...
It is a matter of tradeoff between speed/storage from one side and numerical precision on the other side.

**Comment by grlee77** [comment]
If users have their input in `float32`, they probably chose that over the default python `float` for reasons like @rfezzani mentioned above. I think we should make it easier to avoid unexpected promotions to double precision each time scikit-image functions are called. The work here is in line with prior changes to `img_as_float` to preserve single precision.

In general, optimized libraries taking advantage of SIMD instructions can perform twice as many operations in parallel for single vs. double. In practice most scikit-image functions should see less than a factor two improvement in run time. A few comparisons are given in #5204, for example.

**Comment by FirefoxMetzger** [comment]
@rfezzani I totally get the argument in a GPU computing context, as a GPU tends to have many more single-precision FPUs than double-precision FPUs (if any). This also makes storing in FP64 format a little undesirable. In fact, in this line of thinking it is probably much better to add FP16 support or consider the new and shiny bfloat16 since this puts us into the realm of tensor cores. Ampere adds FP32 (and FP64 support in their A100 line), but processing happens with "reduced precision" (me thinks bfloat16).  On the other hand, as far as I am aware, extending to the GPU isn't on the roadmap for skimage, so I'm hesitant to give this argument a lot of weight.

@grlee77 I also agree with the SIMD argument, i.e., that one can fit more numbers into the register and get a proportional speedup. I'm actually a big fan of SIMD, but this is again not something that, to my knowledge, skimage doesn't explicitly aim to support; mainly for reasons of maintainability. This turns SIMD into a bit of a painful topic. Where we rely on numpy for speed, it can help; however, we opt for cython implementations to speed things up. For cython, there is some implicit support whenever the function happens to be written in such a way that the compiler can auto-vectorize it. If it does happen, we can indeed get a nice boost from FP32, but making it happen in a predictable way in the current codebase is a PITA. In some sense, moving towards cython means moving away from SIMD; at least at the moment.

The storage argument holds in a CPU. I'm actually a little curious if storage is an issue people face. Does anyone have a good scenario? For me, the only time I can't fit everything into main memory is in our deep-learning work. However, in this scenario, the main memory is more of a staging area for the GPU anyway (although pre-processing happens there) and as long as it provides a large enough buffer to feed the GPU it's no problem. It hasn't for me (yet).

Also, please don't misunderstand me. I'm totally in favor of returning FP32 for an FP32 input. Adding FP32 processing definitely doesn't hurt either. I'm mainly musing about the implications on computation time, because - in my mind - CPUs only have double-precision FPUs, and FP32 vs FP64 makes no difference to processing except for being more cache-friendly (twice as many numbers per cache line).

**Comment by rfezzani** [comment]
> ... I'm totally in favor of returning FP32 for an FP32 input. Adding FP32 processing definitely doesn't hurt either...

That's I think the main concern of this issue.

Even if GPU/TPU/SIMD are not targets for skimage project, it may be for some of its users. So we need to make things consistent. I mean that we need to ease the integration of skimage in such contexts.

I think storage argument holds anyway. Even if in most of the use cases memory is not an issue, it is not excluded that it can happen in some scenarii with limited hardware capacities (robotics...). And from the ecolo/green perspective, storage means energy: the more we can save, the better it is ;)

Moreover, @grlee77 made some performance measures in #5204, and the gain obtained using FP32 vs FP64 is absolutely not negligible (~35% for TVL1 registration for example).

Implementing FP32 support being not that painful, I think that we must do the effort.

**Comment by jni** [comment]
Everything @rfezzani said! LOL Additionally, skimage *does* rely on NumPy for many operations, and they are making a significant push on SIMD. Cython SIMD support may also improve in the future. And we may one day start using Numba, which might get even better at it. etc etc. So, performance wise, I think the argument is compelling, and again, as @rfezzani said, for minimum effort.

Finally, the storage issue is actually really huge in biology and 3D volumes. Multi-GB volumes are common. In those scenarios, float32 vs float64 can make the difference between fitting your data or having to come up with clever chunking schemes. In other words, this is a rather cheap way to double our users' RAM! Certainly if someone offered to convert my 16GB laptop to 32GB I would not complain! 😂 

@grlee77 

> I am generally in favor of preserving the floating point precision of the input (although usually promoting float16 to float32). 

I like this. I don't think of dtype as an antipattern though, but rather something complementary — preserve the input, but allow the user to override that. Same as e.g. np.copy.

**Comment by grlee77** [comment]
> >    I am generally in favor of preserving the floating point precision of the input (although usually promoting float16 to float32).

> I like this. I don't think of dtype as an antipattern though, but rather something complementary — preserve the input, but allow the user to override that. Same as e.g. np.copy.

I am not strongly against `dtype` as a kwarg, but would like it to default to `None` with automatic preservation of single vs. double as the default. The main downside with `dtype=None` in the function signature is that users may try to provide an incompatible dtype such as `uint8`, but we can always easily check that the provided dtype is floating point via `np.dtype(dtype).kind == 'f'`.

**Comment by grlee77** [comment]
Great discussion, all. It sounds like everyone is on the same page on expanding float32 dtype support.

What will be most useful for me to proceed with a few additional PRs is reviewing a couple of API related things like the `_float_dtype` helper proposed in #5219 and deciding if we should also add a `dtype` argument by default to all functions supporting float32. 

Secondly, in a handful of cases such as optical flow where there is already a `dtype` argument, but the default is `dtype=np.float32`, should we deprecate that and change the default to `None` as discussed in the prior comment?

It will be helpful to settle these things first before making additional PRs in this area, so that effort in changing things later can be minimized.

**Comment by FirefoxMetzger** [comment]
@grlee77 @jni Could you elaborate on the benefit of adding a `dtype` argument? To me, it sounds equivalent to `some_func(image.astype(dtype, copy=False), ...)` which seems more transparent and easier to maintain compared to the adding `dtype` as an argument.

**Comment by jni** [comment]
> more transparent and easier to maintain compared to the adding dtype as an argument.

Well, I agree that it's only syntactic sugar, but imho inferring from input is *less* transparent in the sense that you have to know that the output copies the dtype of the input. And for functions with multiple inputs you have to guess the promotion rules. etc. etc. `dtype=whatever` makes the intention very clear, I think!

**Comment by grlee77** [comment]
Another discussion point @FirefoxMetzger brought up in #5219

> In general, I wonder if functions should return the same dtype as the input image, or if they should return the dtype in which the processing was done. E.g., if the input is FP16, should it return FP32 or FP16?

Right, for example we cannot easily use float16 computations in our Cython code or in code involving FFTs or PyWavelets, etc. that don't offer 16-bit implementations. We can compute in float32 and cast back to float16 at the end, though. 

Should we generally always cast floating point inputs back to the user-provided types? I can see where this could potentially be useful, but it also might lead to incorrect assumptions about the actual precision of the operations being applied. (e.g. using float128 inputs wouldn't actually give better precision results than float64, but users may not realize this).

**Comment by FirefoxMetzger** [comment]
Here is a thought:

In an ideal world, the user chooses their datatypes for a reason. In reality, most users don't but it is probably worthwhile for us to assume this is the case anyway.  In my mind, this has two consequences:

1.  Computation should generally happen in a `dtype` that is at least as precise as the user's input and preferably in an even larger type if the underlying hardware supports that at the same speed to keep the error minimal. Nobody will complain if the precision is higher than expected - esp. at the same speed -, but if it is lower we can get into a terrible mess. If we demote the `dtype`, because we (or the hardware) can't support it otherwise, we should probably display a warning message to make that known. Given @grlee77 's helper function in #5219 this should be doable in a single place.

2. If we promote the `dtype`, because we can't support it otherwise, then we need to consider that the user likely chose a smaller type because of memory, or because the hardware has an accelerator for that type that we don't support (yet :) ). We can't really do much about the latter without updating our code, but we probably want to do something about the former and at least accumulate results in the same `dtype` as the user's input. Not sure about storing intermediate/temporary variables, but I guess the same thought applies. From that angle, I think it makes sense to return the same `dtype` as the input image, and if there are multiple with the same importance to resolve it by either (1) choosing the one with the highest precision or (2) choosing the one with the most elements and do so consistently. In that special case, it may make sense to think about a `accumulation_dtype` variable that allows manual tuning of the behavior, but I don't know how relevant such a niche thing will be for us.

**Comment by grlee77** [comment]
Deleted the discussion, but leaving this closed as resolved as of v0.19.0 where the majority of functions have float32 support. 

Please open a new issue for any isolated remaining functions founds to have missing float32 support
